### ISSUE TITLE
[CT-377] Configure `profiles_dir` with a config file

### ISSUE THREAD

**Issue by dbrtly** [dbt-labs/dbt-core#4885]
### Is there an existing feature request for this?

- [X] I have searched the existing issues

### Describe the Feature

As per [the documentation](https://docs.getdbt.com/dbt-cli/configure-your-profile#advanced-customizing-a-profile-directory) while this works:

```
$ dbt run --profiles-dir path/to/directory
```
and so does 
```
$ export DBT_PROFILES_DIR=path/to/directory
```

I would prefer to set it in dbt_project.yml and forget about it.

```
[profiles-path]: directorypath
```

### Describe alternatives you've considered

.env file in repo 

### Who will this benefit?

cli users

### Are you interested in contributing this feature?

yes

**Comment by iknox-fa** [comment]
Hi @dbrtly and thanks for the feature request!  I'll add this as a backlog ticket and we'll get it prioritized as soon as it makes sense.  In the interim, I've generally recommended `.env` files for these sorts of use cases in the past.

**Comment by lostmygithubaccount** [comment]
removing stale and clearing assignee, this sounds useful. perhaps we can pick this up for 1.4?

**Comment by jtcohen6** [comment]
Sounds a lot like the changes we made to support `target-path` and `log-path` (https://github.com/dbt-labs/dbt-core/pull/5402). For consistency with those and other "global"/runtime configs, the precedence order would want to be:
1. CLI flag: `--profiles-dir`
2. Env var: `DBT_PROFILES_DIR`
3. Project config: `profiles-dir:` in `dbt_project.yml`

This is also related to @dbeatty10's proposed change, to look for `profiles.yml` in IF a custom `--profiles-dir` is not set (https://github.com/dbt-labs/dbt-core/issues/5411, https://github.com/dbt-labs/dbt-core/pull/5717). Taking that into account, the updated precedence order would actually look like:

1. `--profiles-dir` command-line argument
2. `DBT_PROFILES_DIR` environment variable
3. `profiles-dir` property of `dbt_project.yml` NEW
4. current working directory of `dbt_project.yml` NEW
5. HOME directory of the user (i.e. `~/.dbt/`)

**Comment by dbrtly** [comment]
:) yes please


Thanks,
Daniel
________________________________

**Comment by jtcohen6** [comment]
**Update:** I realized that this might be very tricky because of the order in which dbt parses each config at run initialization. I think we need to read and locate `profiles.yml` first, in order to set `UserConfig`, before we start reading `dbt_project.yml`. In which case, this could be much trickier to implement.

We have been discussing moving `UserConfig` out of `profiles.yml` and into a dedicated `.env` / config file, which would also make sense.

**Comment by dbeatty10** [comment]
Agreed that the implementation would be quite involved!

Oooh, I like the idea of separating the current [`profiles.yml`](https://docs.getdbt.com/reference/profiles.yml) file into separate configuration files 🤩:
1. user configuration
    - `config` key in current [`profiles.yml`](https://docs.getdbt.com/reference/profiles.yml)
    - maybe named something like `user_config.yml`
1. connection configuration
    - `<profile-name>` keys in current [`profiles.yml`](https://docs.getdbt.com/reference/profiles.yml)
    - probably still named `profiles.yml`

@dbrtly how would you imagine using the proposed `profiles-dir` property of `dbt_project.yml`?

Are you aiming to specify the **user** configuration for the dbt project? Or the **database connection** configuration? Or both?

If you have multiple dbt projects, would you intend to point them all at a single shared `profiles.yml` file? Or would you point each of them at their own unique `profiles.yml` files (one per dbt project)?

If each project would have its own `profiles.yml` file, where would you put it? In the same directory as `dbt_project.yml`?

**Comment by dbrtly** [comment]
My initial beef was annoyance at setting $DBT_PROFILES_DIR over and over. These days, I’m also using direnv to run `gcloud auth login …` so direnv is doing more than one thing. 

If dbt looked for profiles.yml in the same directory as the dbt_project.yml file, I would be happy with that. 

If you are going to change this, my perspective is that there should be an option to use a pyproject.toml file for compatibility with the wider modern Python ecosystem. 

Something like:

```
[tools.dbt.config]
# https://docs.getdbt.com/reference/global-configs

send_anonymous_usage_stats = False
use_colors = True
partial_parse = False
printer_width = 90
write_json = True
warn_error_include= all 
log_format = json
debug = False
version_check = False
fail_fast = False
use_experimental_parser = False
static_parser = False

[tools.dbt.targets.dev]
# https://docs.getdbt.com/reference/warehouse-setups/bigquery-setup

type = bigquery
method = oauth
location = “US” # Optional, one of US or EU, or a regional location
# project = use $GOOGLE_PROJECT_ID
# actively discourage hardcoded secret in config

```


Github Actions treats Google project ID as a secret. It never prints it to the logs (always uses a dummy string). I find it a code smell that dbt encourages users to handle the ID with less care.

**Comment by Bilbottom** [comment]
Commenting to add my support for using the `pyproject.toml` file for project configuration 🚀 

Like @dbrtly mentioned, the `pyproject.toml` file is becoming the standard way to configure project settings in the Python ecosystem

I'm less opinioned on how much stuff to put into `pyproject.toml` compared to `dbt_project.yml`, but at the very minimum I would love to be able to specify the `project_dir` and `profiles_dir` so that I can avoid setting `$DBT_PROFILES_DIR` and `$DBT_PROFILES_DIR` each time I want to run dbt (same gripe as dbrtly!)

Relating this to the precedence order mentioned before, I'd tweak item 3 to the `pyproject.toml` file:

1. `--profiles-dir` command-line argument
2. `DBT_PROFILES_DIR` environment variable
3. `profiles_dir` property of `pyproject.toml` <- NEW
4. current working directory of `dbt_project.yml`
5. `HOME` directory of the user (i.e. `~/.dbt/`)

(I think 4 has changed recently to just the current working directory, which is also more intuitive)

This is already supported by [the SQLFluff package's dbt-templater plugin](https://docs.sqlfluff.com/en/stable/configuration.html#dbt-templater):

```toml
[sqlfluff:templater:dbt]
project_dir = <relative or absolute path to dbt_project directory>
profiles_dir = <relative or absolute path to the directory that contains the profiles.yml file>
profile = <dbt profile>
target = <dbt target>
```

**Comment by Bilbottom** [comment]
> This issue has been marked as Stale because it has been open for 180 days with no activity. If you would like the issue to remain open, please comment on the issue or else it will be closed in 7 days.

Commenting to keep this alive

**Comment by dbeatty10** [comment]
Labeling this as `triage` for us to choose between one of the following approaches for this request:
1. Support a `pyproject.toml` file to read dbt project config
1. Add a `profiles-dir` property to `dbt_project.yml`
1. Status quo / do nothing
    - 1. Encourage using an `.env` or `.envrc` ([`direnv`](https://direnv.net/)) file with `DBT_PROFILES_DIR` specified

**Comment by dbeatty10** [comment]
Currently, the parent directory for `profiles.yml` is determined using the following precedence [[1]](https://docs.getdbt.com/docs/core/connect-data-platform/connection-profiles#advanced-customizing-a-profile-directory):

1. `--profiles-dir` CLI flag
1. `DBT_PROFILES_DIR` environment variable
1. current working directory (conventional file location 1)
1. `~/.dbt/ directory` (conventional file location 2)

@jtcohen6 and I got a chance to discuss this briefly today. We decided a couple things:
- Our location for specifying dbt configuration is `dbt_project.yml`, so we're not planning to add an additional file format (i.e., `pyproject.toml`)
- Given that we already have a CLI flag, environment variable, and two conventional file locations, it's not a priority for us to add `profiles-dir` property to `dbt_project.yml`

So I'm closing this issue as "not planned".
